### DISCUSSION TITLE
How can I use translations outside of components?

### DISCUSSION THREAD

**Discussion by hqw567** [amannn/next-intl#729] [Q&A]
How to use translation outside of components? I have some special circumstances where I need to use translation outside of the component. Is there a compatible method to use translation outside of the component? Although it is troublesome and terrible, I still need to consider this situation.

**Comment by nishantbhavasar** [comment]
Hello @hqw567,
i have this same type of requirement have you found any solution to use translation outside of the component.

i am working with next.js or i have implement translation both side client side or server side component but i am calling Api's outside my components and need to handle error or other things in multi language
so if you find any solution of this please provide me 

Thank you

**Comment by hqw567** [comment]
@amannn Can you help us out?

**Comment by amannn** [comment]
Please see https://next-intl-docs.vercel.app/blog/translations-outside-of-react-components

**Reply by hqw567** [reply]
If I must use it outside of a component? (I believe there are many others with the same need as me) Could you consider a compromise solution, rather than instructing users on how to avoid the issue?

**Reply by hqw567** [reply]
In my view, a well-designed library should aim to minimize the cognitive load on developers, enabling them to code more intuitively rather than having them spend a significant amount of time learning how to avoid potential pitfalls. The goal of a library should be to simplify common tasks and make them straightforward, not to force developers to memorize and navigate through a minefield of issues.

**Reply by amannn** [reply]
I agree. The blog post I've linked to tries to explain just that, how issues that might not be apparent upfront can be avoided automatically by restricting message consumption to components. It might feel limiting in certain cases but ultimately tries to push for robust code that avoids bugs in the first place.

**Comment by cini9** [comment]
Just to add to the discussion here. I've found a use case for using translations outside of components using yup (validation library for forms). I've been trying to create a schema like so :

```
export const getContactSchema = (t: any) =>
  object({
    firstName: string().label(t('firstName')).required(),
    lastName: string().label(t('lastName')).required(),
    birthDate: date().label(t('birthDate')).required(),
    gender: string<Gender>().label(t('gender')).required(),
  })
```

which uses the translated labels for error messages based on locale. Now this schema is used in multiple forms, and is concatenated with other schema as needed. So to keep my component code clean and DRY, I extract the individual schema into separate files which are not React components and then fetch them from the React components :

```
const contactSchema = useMemo(() => getContactSchema(t), [t])
```

As you can see, I need to pass the t function to the getContactSchema method in order to specify translated labels, as I am unable to use useTranslations nor getTranslations. Besides having to type the t function correctly (which I have been unable to do so far 😢 ), I don't think that this is scalable for larger applications.

Other libraries expose the i18n object so that we can call i18n.t function outside of components and use translations where needed. 

@amannn I understand your approach, but I find that there are also use cases where translations outside of components are really useful. Is there any way you can reconsider allowing translations outside of components ?

**Reply by hqw567** [reply]
```tsx
'use client'
import { useTranslations } from 'next-intl'
import React, { useEffect } from 'react'
import { create } from 'zustand'

type TranslationType = ReturnType<typeof useTranslations>

interface TranslationsStore {
  t?: TranslationType
  setT: (t: TranslationType) => void
}

export const useTranslationsStore = create<TranslationsStore>((set) => ({
  setT: (t) => {
    set({ t })
  },
}))

export const getTranslations = () => useTranslationsStore.getState().t

const TranslationsProvider = ({
  children,
}: Readonly<{
  children: React.ReactNode
}>) => {
  const t = useTranslations()
  useEffect(() => {
    useTranslationsStore.getState().setT(t)
  }, [t])

  return children
}

export default TranslationsProvider
```

Use `TranslationsProvider` to wrap root layout so it can be obtained anywhere, but it is not safe
```ts
const t = getTranslations()
```

**Reply by cini9** [reply]
@hqw567 Thanks a lot for this. I had thought about using a ReactContext (which I figured wouldn't work since I'm trying to obtain the t function outside of components) but I hadn't thought of storing the t function using zustand. I will definitely give it a try. What do you mean that it's not safe though ?

**Reply by cini9** [reply]
I've tried to set this up, but I've come across one major inconvenience, which I have yet to solve. Essentially, since the t function can be undefined (according to typescript), I have to do a if (t) check everywhere I plan to use the t function coming from getTranslations. Any ideas on how this can be fixed ?

**Reply by hqw567** [reply]
export const getTranslations = () => (useTranslationsStore.getState().t as TranslationType)

**Reply by cini9** [reply]
Thanks @hqw567 I will give it a try !

**Reply by ung0v** [reply]
Thanks @hqw567, this is very useful for working with components outside, but I ran into a TypeScript problem with my JSON file:
```json
{
  "Common": {
    "token_expired": "Token expired"
  }
}
```
I could only write `t("token_expired")`, but it should be `t("Common.token_expired")`.
I found a fix:
```typescript
type TranslationType = ReturnType<typeof useTranslations<never>> 👈  added generic type never
```

**Reply by anuj9196** [reply]
I also have a similar use case. Since using react-table, there is an external function `getColumns` which is in a separate file and too large to put in react component because it will be a reusable function. e.g.,
```tsx
function getActions(item: QrCode, onAction: (action: QrAction, item2: QrCode) => void, t: (s: string) => string) {
  const hasPage = qr.category.has_landing_page && qr.dynamic_url_object?.is_page;
  return {
    primary: [
      { key: "download", icon: Download01Icon, title: t("Download"), onClick: () => onAction("download", qr) },
      ...(hasPage ? [{ key: "editpage", icon: FileEditIcon, title: t("Edit Page"), href: `/qr/page/new/${qr.category.slug}?qrid=${qr.qrid}` }] : []),
      { key: "edit",     icon: Edit01Icon,     title: t("Edit"),     href: `/qr/${qr.qrid}` },
    ],
    dropdown: [
      { key: "analytics", icon: Analytics01Icon,                                    label: t("Analytics"), hidden: qr.qr_type !== "dy", href: `/qr/${qr.qrid}` },
      { key: "duplicate",       icon: Copy02Icon,                                         label: t("Duplicate"),        onClick: () => onAction("duplicate", qr) },
      { key: "move_to_folder",  icon: FolderTransferIcon,                                 label: t("Move to Folder"),   onClick: () => onAction("move_to_folder", qr) },
      { key: "toggle",          icon: qr.is_active ? PauseCircleIcon : PlayCircleIcon,    label: qr.is_active ? t("Deactivate") : t("Activate"), separatorBefore: true, onClick: () => onAction("toggle_active", qr) },
      { key: "delete",    icon: Delete01Icon,                                       label: t("Delete"),    variant: "destructive" as const, separatorBefore: true, onClick: () => onAction("delete", qr) },
    ],
  };
}
```

Though I'm passing `t` to the function like

```tsx
const t = useExtracted();
const actions = getActions(item, (action, item2) => {
), t);
```

getting error in console

```
IntlError: MISSING_MESSAGE: Could not resolve `Download` in messages for locale `en`.
    at getFallbackFromErrorAndNotify (initializeConfig-CUsOI8u2.js:96:19)
    at translateBaseFn (initializeConfig-CUsOI8u2.js:129:18)
    at translateFn (initializeConfig-CUsOI8u2.js:161:20)
```

I believe having this feature will solve issues with complex application architecture

**Reply by amannn** [reply]
@anuj9196 Please see https://next-intl.dev/docs/usage/extraction#static-analysis

While it was always recommended for `useTranslations` to not pass `t` to functions, for `useExtracted` it's required.

You can however call `useExtracted` in a hook like `useTableActions()` if you like.
